auth/downscoping/snippets.py

```diff
@@ -0,0 +1,145 @@
+    # Downscoped credentials will have readonly access to the resource.
+    available_permissions = ["inRole:roles/storage.objectViewer"]
+    # Only objects starting with the specified prefix string in the object name
+    # will be allowed read access.
+    availability_expression = (
+        "resource.name.startsWith('projects/_/buckets/{}/objects/{}')".format(
+            bucket_name, object_prefix
+        )
+    )
+    availability_condition = downscoped.AvailabilityCondition(availability_expression)
+    # Define the single access boundary rule using the above properties.
+    rule = downscoped.AccessBoundaryRule(
+        available_resource=available_resource,
+        available_permissions=available_permissions,
+        availability_condition=availability_condition,
+    )
+    # Define the credential access boundary with all the relevants rules.
+    credential_access_boundary = downscoped.CredentialAccessBoundary(rules=[rule])
+    # [END auth_downscoping_rules]
+
+    # [START auth_downscoping_initialize_downscoped_cred]
+    # Retrieve the source credentials via ADC.
+    source_credentials, _ = google.auth.default()
+    if source_credentials.requires_scopes:
+        source_credentials = source_credentials.with_scopes(
+            ["https://www.googleapis.com/auth/cloud-platform"]
+        )
+
+    # Create the downscoped credentials.
+    downscoped_credentials = downscoped.Credentials(
+        source_credentials=source_credentials,
+        credential_access_boundary=credential_access_boundary,
+    )
+
+    # Refresh the tokens.
+    downscoped_credentials.refresh(requests.Request())
+
+    # These values will need to be passed to the Token Consumer.
+    access_token = downscoped_credentials.token
+    expiry = downscoped_credentials.expiry
+    # [END auth_downscoping_initialize_downscoped_cred]
+    return (access_token, expiry)
+
+
+# [END auth_downscoping_token_broker]
+
+
+# [START auth_downscoping_token_consumer]
+def token_consumer(bucket_name, object_name):
+    """Tests token consumer readonly access to the specified object.
+
+    Args:
+        bucket_name (str): The name of the cloud storage bucket.
+        object_name (str): The name of the object in the storage bucket to
+            read.
+    """
+    # Create the OAuth credentials from the downscoped token and pass a
+    # refresh handler to handle token expiration. Passing the original
+    # downscoped token or the expiry here is optional, as the refresh_handler
+    # will generate the downscoped token on demand.
+    def refresh_handler(request, scopes=None):
+        # The common pattern of usage is to have a token broker pass the
+        # downscoped short-lived access tokens to a token consumer via some
+        # secure authenticated channel.
+        # For illustration purposes, we are generating the downscoped token
+        # locally.
+        return get_token_from_broker(bucket_name, object_name[0:3])
```

Comment:
why `object_name[0:3]`?

Reply:
We want to test the ability to limit access to objects with a certain prefix string in the resource bucket:
https://cloud.google.com/iam/docs/downscoping-short-lived-credentials#example-multiple-buckets

This is an arbitrary index I chose. I didn't want to hardcode the string.

Reply:
maybe adding a comment here to say `object_name[0:3]` is used as an object prefix

Reply:
To remove room for confusion, maybe pass it as a named argument `object_prefix=object_name[0:3]` or instantiate a `object_prefix` variable separately?

```py
# Use the first three characters of the object as the prefix.
object_prefix = object_name[0:3] 
return get_token_from_broker(bucket_name, object_prefix)
```

Reply:
I added a comment explaining this and I am now passing a named argument `object_prefix=object_name[0:3]`